Geometry_Engine/Query/GeometryHash.cs

```diff
@@ -0,0 +1,409 @@
+        {
+            translationFactor += (int)TypeTranslationFactor.Ellipse;
+
+            return curve.StartPoint().ToDoubleArray(translationFactor)
+               .Concat(curve.PointAtParameter(0.33).ToDoubleArray(translationFactor))
+               .Concat(curve.PointAtParameter(0.66).ToDoubleArray(translationFactor))
+               .ToArray();
+        }
+
+        /***************************************************/
+
+        [Description("The GeometryHash for a Line is calculated as the GeometryHash of the start and end point of the Line.")]
+        private static double[] GeometryHash(this Line curve, double translationFactor)
+        {
+            translationFactor += (int)TypeTranslationFactor.Line;
+
+            return curve.StartPoint().ToDoubleArray(translationFactor)
+               .Concat(curve.EndPoint().ToDoubleArray(translationFactor))
+               .ToArray();
+        }
+
+        /***************************************************/
+
+        [Description("Moving control points by a translation factor composed by the weights " +
+            "and a subarray of the knot vector. " +
+            "The subarray is made by picking as many elements from the knot vector as the curve degree value.")]
+        private static double[] GeometryHash(this NurbsCurve curve, double translationFactor)
+        {
+            int curveDegree = curve.Degree();
+
+            if (curveDegree == 1)
+                return BH.Engine.Geometry.Create.Polyline(curve.ControlPoints).GeometryHash(translationFactor);
+
+            translationFactor += (int)TypeTranslationFactor.NurbsCurve;
+
+            List<double> concatenated = new List<double>();
+            for (int i = 0; i < curve.ControlPoints.Count(); i++)
+            {
+                double sum = curve.Knots.GetRange(i, curveDegree).Sum();
+                double[] doubles = curve.ControlPoints[i].GeometryHash(sum + curve.Weights[i] + translationFactor);
+                concatenated.AddRange(doubles);
+            }
+
+            return concatenated.ToArray();
+
+            // Simpler & faster but potentially less reliable:
+            // return curve.Knots.Take(pointWeights.Count())
+            //    .Zip(curve.ControlPoints, (k, p) => new double[]{ p.X + k, p.Y + k, p.Z + k })
+            //    .SelectMany(arr => arr).ToArray();
+        }
+
+
+        /***************************************************/
+        /****  Surfaces                                 ****/
+        /***************************************************/
+
+        private static double[] GeometryHash(this ISurface obj, double translationFactor)
+        {
+            return GeometryHash(obj as dynamic, translationFactor);
+        }
+
+        /***************************************************/
+
+        [Description("The GeometryHash for a Line is calculated as the GeometryHash of the External and Internal boundary curves, then concatenated.")]
```

Comment:
```suggestion
        [Description("The GeometryHash for a PlanarSurface is calculated as the GeometryHash of the External and Internal boundary curves, then concatenated.")]
```

Good to check through the types mentioned in the descs, or to simply not mention them at all, as inplicit from the target type for the method.

Reply:
Sorry I did copy paste and missed some types.